astropy/units/core.py

```diff
@@ -2215,17 +2212,15 @@ def _hash(self) -> int:
         return hash((self.name, self._represents))
 
     @classmethod
-    def _from_physical_type_id(cls, physical_type_id):
+    def _from_physical_type_id(
+        cls, physical_type_id: tuple[tuple[str, UnitPower], ...]
+    ) -> UnitBase:
+        if len(physical_type_id) == 1 and physical_type_id[0][1] == 1:
+            return cls(physical_type_id[0][0])
         # get string bases and powers from the ID tuple
         bases = [cls(base) for base, _ in physical_type_id]
         powers = [power for _, power in physical_type_id]
-
-        if len(physical_type_id) == 1 and powers[0] == 1:
-            unit = bases[0]
-        else:
-            unit = CompositeUnit(1, bases, powers, _error_check=False)
```

Comment:
Mypy didn't like this code because it inferred from the `if`-block that the type of `unit` was `cls`, i.e. `Unit` or a subclass of `Unit`, but in the `else`-block the code assigned a `CompositeUnit` to `unit`. The simplest way to avoid having Mypy infer a needlessly narrow type for `unit` was to eliminate `unit` entirely.